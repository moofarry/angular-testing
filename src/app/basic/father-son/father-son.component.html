<h5>Hijo</h5>

<code>{{client | json}}</code>
<br><br>

<button *ngIf="client" class="btn btn-outline-primary" (click)="onChange(5)" data-test="btn-change-id">Cambiar
  ID:</button>
<button *ngIf="client" (click)="onDelete()" class="btn btn-outline-danger" data-test="btn-delete">Borrar
  Cliente:</button>