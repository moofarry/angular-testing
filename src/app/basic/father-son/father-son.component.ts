import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Client } from '../interfaces';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.css'],
})
export class FatherSonComponent {
  @Input() client?: Client;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onClientUpdate = new EventEmitter<Client>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange(id: number) {
    if (!this.client) return;
    this.client = {
      ...this.client,
      id,
    };
    this.onClientUpdate.emit({ ...this.client });
    // this.client.id = newId;
  }
}
