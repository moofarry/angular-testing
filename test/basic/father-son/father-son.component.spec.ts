import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [FatherSonComponent],
    });
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  test('should be match with snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('should not render btn if clients doestn exist', () => {
    const btn = compiled.querySelectorAll('button');
    expect(btn.length).toBe(0);
  });
  test('should  render two btn if clients  exist', () => {
    component.client = { id: 1, name: 'test' };
    fixture.detectChanges();
    const btn = compiled.querySelectorAll('button');
    expect(btn.length).toBe(2);
  });
  test('should be match with snapshow if exist client', () => {
    component.client = { id: 1, name: 'test' };
    fixture.detectChanges();
    expect(compiled).toMatchSnapshot();
  });

  test('should be emit event on delete', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onDeleteClient, 'emit');

    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    btnDelete?.dispatchEvent(new Event('click'));
    expect(component.onDeleteClient.emit).toHaveBeenCalled();
  });

  test('debe de emitir onClientUpdated con el boton de "Cambiar ID"', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onClientUpdate, 'emit');

    const btnChangeId = compiled.querySelector('[data-test=btn-change-id]');
    btnChangeId?.dispatchEvent(new Event('click'));

    expect(component.onClientUpdate.emit).toHaveBeenCalledWith({
      id: 5,
      name: 'Juan',
    });
  });

  test('should emit onchangeClient with ID if clientExist', () => {
    jest.spyOn(component.onClientUpdate, 'emit');
    component.onChange(5);

    expect(component.onClientUpdate.emit).not.toHaveBeenCalled();

    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    component.onChange(1);
    expect(component.onClientUpdate.emit).toHaveBeenCalledWith({
      id: 1,
      name: 'Juan',
    });
  });
});
