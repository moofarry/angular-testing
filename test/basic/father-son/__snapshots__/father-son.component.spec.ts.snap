// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent should be match with snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;

exports[`FatherSonComponent should be match with snapshow if exist client 1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
    {
  "id": 1,
  "name": "test"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary"
    data-test="btn-change-id"
  >
    Cambiar ID:
  </button>
  
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
    Borrar Cliente:
  </button>
  
</div>
`;
